src/gotcha_dl.c

```diff
@@ -5,7 +5,137 @@
+                if (info->dlpi_addr == 0) base_addr = info->dlpi_phdr[i].p_offset;
+                ElfW(Addr) start_addr = base_addr + info->dlpi_phdr[i].p_vaddr;
+                ElfW(Addr) end_addr = start_addr + info->dlpi_phdr[i].p_memsz;
+                if (addrs->lookup_addr >= start_addr && addrs->lookup_addr <= end_addr) {
+                    name = info->dlpi_name;
+                    break;
+                }
+            }
+        }
+        if (name) {
+            struct link_map *current = addrs->lmap;
+            while (current) {
+                if (strcmp(current->l_name, name) == 0) {
+                    addrs->found = 1;
+                    addrs->lmap = current;
+                    break;
+                }
+                current = current->l_next;
+            }
+        }
+    }
+    return 0;
+}
+/**
+ * This function looks for all symbols defined in rela or rel.
+ *
+ * @param name, symbol name
+ * @param lmap, library where we search for symbol.
+ * @return symbol pointer
+ */
+int dl_lookup_symbol_x(const char *name, struct link_map *lmap, void** symbol) {
+    long result;
+    INIT_DYNAMIC(lmap)
+    if (gnu_hash) {
+        debug_printf(3, "Checking GNU hash for %s in %s\n", name, LIB_NAME(lmap));
+        struct gnu_hash_header * header;
+        if (is_rela) header = (struct gnu_hash_header *) (gnu_hash);
+        else header = (struct gnu_hash_header *) (lmap->l_addr + gnu_hash);
+        result = lookup_gnu_hash_symbol(name, symtab, strtab, header);
+        if (result != -1) {
+            if (GOTCHA_CHECK_VISIBILITY(symtab[result])) {
+                *symbol = (void *) (symtab[result].st_value + lmap->l_addr);
+                return 0;
+            } else {
+                debug_printf(3, "Symbol %s found but not exported in %s\n", name, LIB_NAME(lmap));
+            }
+        } else {
+            debug_printf(3, "%s not found in %s\n", name, LIB_NAME(lmap));
+        }
+    } else {
+        debug_printf(3, "Library %s does not export symbols\n", LIB_NAME(lmap));
+    }
+    return -1;
+}
+
+/**
+ * Implement the logic of _dl_sym
+ * 1. find the caller library using the program headers.
+ * 2. find the first library which has the symbol for RTLD_DEFAULT
+ * 3. find the second library which has the symbol for RTLD_NEXT
+ * @param handle, handle for dl operation
+ * @param name, name of the symbol
+ * @param who, the virtual address of the caller
+ * @return symbol pointer
+ */
+
+static void * gotchas_dlsym_rtld_next_lookup(void *handle, const char *name, void *who) {
+    ElfW(Addr) caller = (ElfW(Addr)) who;
+    /* Iterative over the library headers and find the caller
+     * the address of the caller is set in addrs->library_laddr
+     **/
+    struct Addrs addrs;
+    addrs.found = 0;
+    addrs.lookup_addr = caller;
+    addrs.lmap = _r_debug.r_map;
+    void* symbol;
+    dl_iterate_phdr(lib_header_callback, &addrs);
+    if (addrs.lmap->l_next) {
```

Comment:
Also, this 'if' statement is unnecessary. If addrs.lmap->l_ next is NULL, then the below 'for' loop will just not iterate over anything.

Reply:
fixed